sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -114,7 +114,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * @since 1.4.0
    */
   def option(key: String, value: String): DataFrameReader = {
-    this.extraOptions += (key -> value)
```

Comment:
why `+=` doesn't work?

Reply:
Yes. It doesn't work, because `CaseInsensitiveMap.+` operator is designed to return `Map` instead of `CaseInsensitiveMap`.

Reply:
In this PR, I added `++` operator in order to avoid to change the existing API and also to minimize the change. `++` is required because it's used in any way.

Reply:
but `++` is inefficient: we create an intermediate map for no reason.

I don't think changing the return type of `+` is a breaking change. `CaseInsensitiveMap` is a private class. When we pass it to DS V1, it becomes `Map[String, String]`.

Reply:
Got it. I can change that, too. BTW, `++` operation is still required, @cloud-fan .

Reply:
Also, it will break many other code path, @cloud-fan .

Reply:
Currently, `CaseInsensitiveMap.+` doesn't support multiple input.
```
[error]     too many arguments (2) for method +: (kv: (String, B1))org.apache.spark.sql.catalyst.util.CaseInsensitiveMap[B1]
[error]     expansion: this.extraOptions = this.extraOptions.+("url".$minus$greater(url), "dbtable".$minus$greater(table))
[error]     this.extraOptions += ("url" -> url, "dbtable" -> table)
```

I'll keep `++` operator + `Map` combination for that code path.